src/server/game/Movement/MotionMaster.cpp

```diff
@@ -56,144 +59,465 @@ inline bool IsStatic(MovementGenerator* movement)
 
-    _cleanFlag |= MOTIONMMASTER_CLEANFLAG_UPDATE;
-    if (!top()->Update(_owner, diff))
+    MovementGenerator* top = GetCurrentMovementGenerator();
+    if (HasFlag(MOTIONMASTER_FLAG_STATIC_INITIALIZATION_PENDING) && IsStatic(top))
     {
-        _cleanFlag &= ~MOTIONMMASTER_CLEANFLAG_UPDATE;
-        MovementExpired();
+        RemoveFlag(MOTIONMASTER_FLAG_STATIC_INITIALIZATION_PENDING);
+        top->Initialize(_owner);
+    }
+    if (top->HasFlag(MOVEMENTGENERATOR_FLAG_INITIALIZATION_PENDING))
+        top->Initialize(_owner);
+    if (top->HasFlag(MOVEMENTGENERATOR_FLAG_DEACTIVATED))
+        top->Reset(_owner);
+
+    ASSERT(!top->HasFlag(MOVEMENTGENERATOR_FLAG_INITIALIZATION_PENDING | MOVEMENTGENERATOR_FLAG_DEACTIVATED), "MotionMaster:Update: update called on an uninitialized top! (%s) (type: %u, flags: %u)", _owner->GetGUID().ToString().c_str(), top->GetMovementGeneratorType(), top->Flags);
+
+    if (!top->Update(_owner, diff))
+    {
+        ASSERT(top == GetCurrentMovementGenerator(), "MotionMaster::Update: top was modified while updating! (%s)", _owner->GetGUID().ToString().c_str());
+
+        // Since all the actions that modify any slot are delayed, this movement is guaranteed to be top
+        Pop(true, true); // Natural, and only, call to MovementInform
     }
-    else
-        _cleanFlag &= ~MOTIONMMASTER_CLEANFLAG_UPDATE;
 
-    if (!_expireList.empty())
-        ClearExpireList();
+    RemoveFlag(MOTIONMASTER_FLAG_UPDATE);
+
+    while (!_delayedActions.empty())
+    {
+        _delayedActions.front().Resolve();
+        _delayedActions.pop_front();
+    }
 }
 
-void MotionMaster::Clear(bool reset /*= true*/)
+void MotionMaster::Add(MovementGenerator* movement, MovementSlot slot/* = MOTION_SLOT_ACTIVE*/)
 {
-    if (_cleanFlag & MOTIONMMASTER_CLEANFLAG_UPDATE)
+    if (!movement)
+        return;
+
+    if (IsInvalidMovementSlot(slot))
     {
-        if (reset)
-            _cleanFlag |= MOTIONMMASTER_CLEANFLAG_RESET;
-        else
-            _cleanFlag &= ~MOTIONMMASTER_CLEANFLAG_RESET;
-        DelayedClean();
+        delete movement;
+        return;
+    }
+
+    if (HasFlag(MOTIONMASTER_FLAG_UPDATE))
+    {
+        std::function<void()> action = [this, movement, slot]() {
+            Add(movement, slot);
+        };
+        _delayedActions.emplace_back(std::move(action), MOTIONMASTER_DELAYED_ADD);
     }
     else
-        DirectClean(reset);
+        DirectAdd(movement, slot);
 }
 
-void MotionMaster::Clear(MovementSlot slot)
+void MotionMaster::Remove(MovementGenerator* movement, MovementSlot slot/* = MOTION_SLOT_ACTIVE*/)
 {
-    if (empty() || IsInvalidMovementSlot(slot))
+    if (!movement || IsInvalidMovementSlot(slot))
         return;
 
-    if (_cleanFlag & MOTIONMMASTER_CLEANFLAG_UPDATE)
-        DelayedClean(slot);
-    else
-        DirectClean(slot);
+    if (HasFlag(MOTIONMASTER_FLAG_UPDATE))
+    {
+        std::function<void()> action = [this, movement, slot]() {
```

Comment:
codestyle, `{` placement (applies to all std::function objects you added for _delayedActions)

Reply:
tru tru